### ISSUE TITLE
Report per month/session

### ISSUE THREAD

**Issue by pierrekil** [openremote/openremote#1323]
How can we best cover these type use cases?:
- I want to see the energy consumed for that month, so that I can invoice my customers
- I want to see the amount of kWh's charged per charge session so I can invoice the car owner
- I want to see it on a dashboard, export the data either manually (CSV?) or link it to an invoicing application

**Comment by pankalog** [comment]
@pierrekil I stumbled on this issue and I had an idea:

I have been messing around with [GraphQL](https://graphql.org/) recently. It is meant to be a query language for interacting with APIs, but it can easily be used as a way to interact with data itself (in fact, I find it a lot more intuitive). GraphQL allows searching, sorting, filtering, similarity search, aggregate queries, statistics like count-sum-average-etc, boolean algebra, persistent queries, etc. Along with GraphQL, there exists [GraphiQL](https://github.com/graphql/graphiql?tab=readme-ov-file#graphiql), a nice, open-source user interface for interacting with a GraphQL API. 

It is an easier and more "clean" way of interacting with the asset datapoint table without interacting with the database directly. The scope of this is very big, so I think that if we want to keep it "open-ended", any user would appreciate a GraphQL interface to access data from a database table, since it allows for the biggest amount of adjustment. It would be a sort-of logical equivalent to using Groovy rules, it allows that extra granularity on top of the when-then rules. For accessing historical data, it would allow users to make more specialized and granular queries to get the data that will generate their report. And on top of that, it would be a fire-and-forget query, that can be cached and executed whenever requested by the user, meaning that there is no need to edit the query.

The GraphQL interface could be paired with (a) widget(s), that will execute that GraphQL query and print the data as either a table or a graph or a number, if the output is two dimensional, one dimensional, or a single value. It could also allow as an export to CSV, xlsx, etc. 

If you want to I can find a functional demo of this with some test data.

**Comment by Hackerberg43** [comment]
@pierrekil Could you elleborate how a "session" would be defined here? I can imagine the data being generated by a rule, but would you imagine tagging it ? A session would be defined by two moments in time, this would require a new data format?

I'm currently working on a widget that uses the datatpoint interval API for getting avg/min/max/delta, but that's only for regular periods.

![Image](https://github.com/user-attachments/assets/c7f44b79-2a4f-4a43-85b4-dba03ac71211)

**Comment by pierrekil** [comment]
Hi @Hackerberg43 we were indeed thinking about a new attribute value type. The bare minimum would be to have it able to store a start and end time. Further logic would be required to link the sessions e.g. to a user, or a energy change, or miles driven. I can imagine that you use this attribute to query the DB to show e.g. max power per 'session' or 'energy consumption per session'. Is that what you have in mind?

**Comment by Hackerberg43** [comment]
Right, then we are on the same page, I will keep this in mind while working on reports, but will limit it to regular time intervals for now.
